tools/python/boutiques/validator.py

```diff
@@ -0,0 +1,221 @@
+        print("JSON Validation error (Boutiques validation not yet performed)")
+        print("Error: {}".format(e))
+        return -1
+
+    # Helper get-function
+    safeGet   = lambda desc, sec, targ: [item[targ] for item in desc[sec]
+                                         if list(item.keys()).count(targ)]
+    inputGet  = lambda s: safeGet(descriptor, "inputs", s)
+    outputGet = lambda s: safeGet(descriptor, "output-files", s)
+    groupGet  = lambda s: safeGet(descriptor, "groups", s)
+    inById    = lambda i: descriptor["inputs"][inputGet("id").index(i)] if i in inputGet("id") else {}
+
+    # Begin looking at Boutiques-specific failures
+    errors = []
+
+    clkeys = inputGet("value-key") + outputGet("value-key")
+    configFileTemplates = outputGet("file-template") + outputGet("path-template")
+    cmdline = descriptor["command-line"]
+
+    # Verify that all command-line key appear in the command-line
+    msg_template = "   KeyError: \"{}\" not in command-line or file template"
+    errors += [msg_template.format(k.strip("[]"))
+               for k in clkeys
+               if (cmdline.count(k) + " ".join(configFileTemplates).count(k)) < 1]
+
+    # Verify that no key contains another key
+    msg_template = "   KeyError: \"{}\" contains \"{}\""
+    errors += [msg_template.format(key.strip("[]"), clkeys[jdx].strip("[]"))
+               for idx, key in enumerate(clkeys)
+               for jdx in range(0, len(clkeys))
+               if clkeys[jdx].strip("[]") in key and key is not clkeys[jdx]]
+
+    # Verify that all Ids are unique
+    inIds, outIds, grpIds = inputGet("id"), outputGet("id"), groupGet("id")
+    allIds = inIds + outIds + grpIds
+    msg_template = "    IdError: \"{}\" is non-unique"
+    for idx, s1 in enumerate(allIds):
+        for jdx, s2 in enumerate(allIds):
+            errors += [msg_template.format(s1)] if s1 == s2 and idx < jdx else []
+
+    # Verify that output files have unique path-templates
+    msg_template = "OutputError: \"{}\" and \"{}\" have the same path-template"
+    for idx, out1 in enumerate(descriptor["output-files"]):
+        for jdx, out2 in enumerate(descriptor["output-files"]):
+            errors += [msg_template.format(out1["id"], out2["id"])] if out1["path-template"] == out2["path-template"] and jdx > idx else []
+
+    # Verify inputs
+    for inp in descriptor["inputs"]:
+
+        # Verify flag-type inputs (have flags, not required, cannot be lists)
+        if inp["type"] == "Flag":
+            msg_template = " InputError: \"{}\" must have a command-line flag"
+            errors += [msg_template.format(inp["id"])] if "command-line-flag" not in inp.keys() else []
+
+            msg_template = " InputError: \"{}\" should not be required"
+            errors += [msg_template.format(inp["id"])] if inp["optional"] is False else []
+
+            # This one is redundant as basic JSON validation catches it
+            msg_template = " InputError: \"{}\" cannot be a list"
```

Comment:
This should now be caught by the schema, could you verify and remove the code if confirmed? Catching that flags aren't required and have a command line flag should be feasible in the schema, by adding to the object in line 326 of the current schema (but let's leave it for another PR). See also #80.